Storage/src/Connection/RetryTrait.php

```diff
@@ -0,0 +1,244 @@
+<?php
+/**
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage\Connection;
+
+/**
+ * Trait which provides helper methods for retry logic.
+ *
+ * @internal
+ */
+trait RetryTrait
```

Comment:
It's not used by any other component/product as of now(_And I don't see that happening in the future too_).
The logic of `retrying all idempotent ops, retrying conditional ops only when precondition is present' etc, is very specific to Storage.

Adding it to `Core` seems like a lot of noise for a package used by multiple products.
I could remove the trait and put all of it in `Connection\Rest` since it is not used in any other class, but this was done to separate the retrying logic from the other parts.

Reply:
As long as all these classes/functions/properties are private or marked `@internal` (so we can refactor them later without breaking BC), we don't need to worry about making this generic until it's something we actually need.

However, the options we expose on the clients would not be able to change, so I suggest we either don't expose them, or do not document them, until we know it's something our customers actually need.